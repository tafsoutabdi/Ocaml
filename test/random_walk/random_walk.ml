open Definitions.MarkovChain
open MarkovChain.RandomWalk
open Utils.Std

let () = Random.self_init ()

let estimate_distribution sampler nb_samples =
  let res = Hashtbl.create 100 in
  let rec loop i =
    if i=nb_samples then res else
    let sample = sampler () in
    let n = try Hashtbl.find res sample with Not_found -> 0 in
    Hashtbl.replace res sample (n+1);
    loop (i+1)
  in loop 0 

let runtest length ?(start=0) nb_samples =
  Format.printf "`random_walk ~length:%d ~start:%d mc_example` renvoie@."
    length start;
  let dist = estimate_distribution (fun ()-> 
    random_walk ~length ~start mc_example |> List.map Char.escaped |> String.concat ""
    ) nb_samples in
  dist 
  |> Hashtbl.to_seq_keys |> List.of_seq |> List.sort compare 
  |> List.iter (fun walk ->
    let n = Hashtbl.find dist walk in
    let proba =  percent_of_frac n nb_samples in
    Format.printf "%s avec probabilit√© %d%%@." walk proba
  );
  Format.printf "@."

let () = runtest 1 5 

let () = runtest 2 100_000

let () = runtest 3 1_000_000 

let () = runtest 4 1_000_000

let () = runtest 5 1_000_000 