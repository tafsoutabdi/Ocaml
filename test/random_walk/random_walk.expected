`random_walk ~length:1 ~start:0 mc_example` renvoie
a avec probabilité 100%

`random_walk ~length:2 ~start:0 mc_example` renvoie
ab avec probabilité 50%
ag avec probabilité 50%

`random_walk ~length:3 ~start:0 mc_example` renvoie
abc avec probabilité 25%
abd avec probabilité 8%
abe avec probabilité 17%
ag avec probabilité 50%

`random_walk ~length:4 ~start:0 mc_example` renvoie
abc avec probabilité 25%
abd avec probabilité 8%
abef avec probabilité 17%
ag avec probabilité 50%

`random_walk ~length:5 ~start:0 mc_example` renvoie
abc avec probabilité 25%
abd avec probabilité 8%
abefb avec probabilité 8%
abefg avec probabilité 8%
ag avec probabilité 50%

